### PR TITLE
Add support for shielded instance

### LINKED ISSUE #776: Add Shielded VM support
/kind feature

**Describe the solution you'd like**
I'd like to be able to modify the default shielded VM options, more specifically I'd like the option to enable secure boot.
 

**Anything else you would like to add:**

### PR DESCRIPTION
**Pull request by eranco74**
/kind feature
/kind api-change

**What this PR does / why we need it**:
The PR adds support for shielded instances

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #776

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

**Release note**:

```release-note
Add shielded instance config  to API
```

### CODE REVIEW

File: api/v1beta1/gcpmachine_types.go [hunk 1]
```diff
@@ -66,6 +66,62 @@ const (
 	IPForwardingDisabled IPForwarding = "Disabled"
 )
 
+type SecureBootPolicy string
+
+const (
+	// SecureBootPolicyEnable enables the secure boot configuration for the GCP machine.
+	SecureBootPolicyEnable SecureBootPolicy = "Enabled"
+	// SecureBootPolicyDisable disables the secure boot configuration for the GCP machine.
+	SecureBootPolicyDisable SecureBootPolicy = "Disable"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:75]: Should this be `Disabled` to match the `Enabled` semantic?

File: api/v1beta1/gcpmachine_types.go [hunk 2]
```diff
@@ -66,6 +66,62 @@ const (
...
+)
+
+type VTPMPolicy string
+
+const (
+	// VTPMPolicyEnable enables the virtualized trusted platform module configuration for the GCP machine.
+	VTPMPolicyEnable VTPMPolicy = "Enabled"
+	// VTPMPolicyDisable disables the virtualized trusted platform module configuration for the GCP machine.
+	VTPMPolicyDisable VTPMPolicy = "Disable"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:84]: I think this should be Disabled too

File: api/v1beta1/gcpmachine_types.go [hunk 3]
```diff
@@ -66,6 +66,62 @@ const (
...
+)
+
+type IntegrityMonitoringPolicy string
+
+const (
+	// IntegrityMonitoringPolicyEnable enables integrity monitoring for the GCP machine.
+	IntegrityMonitoringPolicyEnable IntegrityMonitoringPolicy = "Enabled"
+	// IntegrityMonitoringPolicyDisable disables integrity monitoring for the GCP machine.
+	IntegrityMonitoringPolicyDisable IntegrityMonitoringPolicy = "Disable"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:93]: I think this should also be Disabled

File: api/v1beta1/gcpmachine_types.go [hunk 4]
```diff
@@ -66,6 +66,62 @@ const (
...
+	//+optional
+	SecureBoot SecureBootPolicy `json:"secureBoot,omitempty"`
+
+	// VTPM enable virtualized trusted platform module measurements to create a known good boot integrity policy baseline.
+	// The integrity policy baseline is used for comparison with measurements from subsequent VM boots to determine if anything has changed.
+	// If omitted, the platform chooses a default, which is subject to change over time, currently that default is Enabled.
+	// +kubebuilder:validation:Enum=Enabled;Disabled
+	// +kubebuilder:default=Enabled
+	// +optional
+	VTPM VTPMPolicy `json:"vTPM,omitempty"`
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:113]: Acronyms aren't particularly user friendly, do we want to spell this out as something more explicit?

File: api/v1beta1/gcpmachine_types.go [hunk 5]
```diff
@@ -66,6 +66,62 @@ const (
...
+)
+
+// GCPShieldedInstanceConfig describes the shielded VM configuration of the instance on GCP.
+// Shielded VM configuration allow users to enable and disable Secure Boot, vTPM, and Integrity Monitoring
+type GCPShieldedInstanceConfig struct {
+	// SecureBoot Defines whether the instance should have secure boot enabled.
+	// Secure Boot verify the digital signature of all boot components, and halting the boot process if signature verification fails.
+	// If omitted, the platform chooses a default, which is subject to change over time, currently that default is Disabled.
+	// +kubebuilder:validation:Enum=Enabled;Disabled
+	// +kubebuilder:default=Disabled
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:103]: Setting a default value like this isn't compatible with changing over time, normally the change over time idea is when there is no default set, we should drop one or the other

File: api/v1beta1/gcpmachine_types.go [hunk 6]
```diff
@@ -66,6 +66,62 @@ const (
 	IPForwardingDisabled IPForwarding = "Disabled"
 )
 
+type SecureBootPolicy string
+
+const (
+	// SecureBootPolicyEnable enables the secure boot configuration for the GCP machine.
+	SecureBootPolicyEnable SecureBootPolicy = "Enabled"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:73]:
```suggestion
	SecureBootPolicyEnabled SecureBootPolicy = "Enabled"
```

File: api/v1beta1/gcpmachine_types.go [hunk 7]
```diff
@@ -66,6 +66,62 @@ const (
...
+	// SecureBootPolicyDisable disables the secure boot configuration for the GCP machine.
+	SecureBootPolicyDisable SecureBootPolicy = "Disable"
+)
+
+type VTPMPolicy string
+
+const (
+	// VTPMPolicyEnable enables the virtualized trusted platform module configuration for the GCP machine.
+	VTPMPolicyEnable VTPMPolicy = "Enabled"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:82]:
```suggestion
	VTPMPolicyEnabled VTPMPolicy = "Enabled"
```

File: api/v1beta1/gcpmachine_types.go [hunk 8]
```diff
@@ -66,6 +66,62 @@ const (
...
+	// VTPMPolicyDisable disables the virtualized trusted platform module configuration for the GCP machine.
+	VTPMPolicyDisable VTPMPolicy = "Disable"
+)
+
+type IntegrityMonitoringPolicy string
+
+const (
+	// IntegrityMonitoringPolicyEnable enables integrity monitoring for the GCP machine.
+	IntegrityMonitoringPolicyEnable IntegrityMonitoringPolicy = "Enabled"
```
Comment (JoelSpeed) [inline api/v1beta1/gcpmachine_types.go:91]:
```suggestion
	IntegrityMonitoringPolicyEnabled IntegrityMonitoringPolicy = "Enabled"
```

File: api/v1beta1/gcpmachine_types.go (hunk 1 above)
Comment (eranco74) [reply api/v1beta1/gcpmachine_types.go:75]: Yes, updating

File: api/v1beta1/gcpmachine_types.go (hunk 5 above)
Comment (eranco74) [reply api/v1beta1/gcpmachine_types.go:103]: OK

File: api/v1beta1/gcpmachine_types.go (hunk 4 above)
Comment (eranco74) [reply api/v1beta1/gcpmachine_types.go:113]: Well, it's called `EnableVtpm` in the compute api.

**Review by cpanato — COMMENTED** [review]
/approve
/hold for more reviews
/assign @richardcase

**Comment by richardcase** [comment]
Just a nit (which is just personal preference), so:

/lgtm

File: cloud/scope/machine.go [hunk 9]
```diff
@@ -351,6 +351,21 @@ func (m *MachineScope) InstanceSpec() *compute.Instance {
 	if m.GCPMachine.Spec.IPForwarding != nil && *m.GCPMachine.Spec.IPForwarding == infrav1.IPForwardingDisabled {
 		instance.CanIpForward = false
 	}
+	if m.GCPMachine.Spec.ShieldedInstanceConfig != nil {
+		instance.ShieldedInstanceConfig = &compute.ShieldedInstanceConfig{}
```
[the remark this answers is not in the document]
Comment (JoelSpeed) [reply cloud/scope/machine.go:355]: I like this suggestion FWIW

**Review by JoelSpeed — COMMENTED** [review]
No further feedback from me now, though I do like the suggestion from @richardcase

**Comment by richardcase** [comment]
@eranco74 - appreciate you took the time to look at the nit 🙇 

/lgtm

**Pull request merged**